### PR TITLE
[Ubuntu] Add VCPKG_ROOT variable

### PR DESCRIPTION
**Pull request by erik-bershel**
# Description
Add VCPKG_ROOT variable in additional to VCPKG_INSTALLATION_ROOT

#### Related issue:
[#4241](https://github.com/actions/runner-images-internal/issues/4241)

## Check list
- [x] Related issue / work item is attached
- [x] Tests are written (if applicable)
- [x] Documentation is updated (if applicable)
- [ ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

**Review by ddobranic — DISMISSED** [review]

File: images/linux/scripts/installers/vcpkg.sh [hunk 1]
```diff
@@ -4,16 +4,19 @@
 ##  Desc:  Installs vcpkg
 ################################################################################
 
-# Set env variable for vcpkg
+# Set env variables for vcpkg
 VCPKG_INSTALLATION_ROOT=/usr/local/share/vcpkg
 echo "VCPKG_INSTALLATION_ROOT=${VCPKG_INSTALLATION_ROOT}" | tee -a /etc/environment
 
+VCPKG_ROOT=/usr/local/share/vcpkg
+echo "VCPKG_ROOT=${VCPKG_ROOT}" | tee -a /etc/environment
```
Comment (miketimofeev) [inline images/linux/scripts/installers/vcpkg.sh:12]:
```suggestion
VCPKG_INSTALLATION_ROOT=/usr/local/share/vcpkg
echo "VCPKG_INSTALLATION_ROOT=${VCPKG_INSTALLATION_ROOT}" | tee -a /etc/environment
echo "VCPKG_ROOT=${VCPKG_INSTALLATION_ROOT}" | tee -a /etc/environment
```

File: images/linux/scripts/installers/vcpkg.sh [hunk 2]
```diff
@@ -4,16 +4,19 @@
 ##  Desc:  Installs vcpkg
 ################################################################################
 
-# Set env variable for vcpkg
+# Set env variables for vcpkg
 VCPKG_INSTALLATION_ROOT=/usr/local/share/vcpkg
 echo "VCPKG_INSTALLATION_ROOT=${VCPKG_INSTALLATION_ROOT}" | tee -a /etc/environment
 
+VCPKG_ROOT=/usr/local/share/vcpkg
```
Comment (mikhailkoliada) [inline images/linux/scripts/installers/vcpkg.sh:11]: why define it here and not upper then?

**Review by mikhailkoliada — CHANGES_REQUESTED** [review]

**Review by al-cheb — APPROVED** [review]

**Review by ddobranic — APPROVED** [review]

**Review by miketimofeev — APPROVED** [review]

**Comment by BillyONeal** [comment]
This breaks vcpkg customers who use vcpkg as a submodule. Can you explain why this variable is being set?

**Comment by al-cheb** [comment]
@BillyONeal, we got a request to add this variable https://github.com/actions/runner-images/discussions/6146. @michaelbprice, Could you please provide more context?

**Comment by michaelbprice** [comment]
There are at least 3 techniques being used in the wild to acquire vcpkg.
1) Included as a git submodule, which is then acquired/updated using git's submodule commands.
2) Downloaded via a one-line shell command (as shown in https://devblogs.microsoft.com/cppblog/vcpkg-environment-activation-in-visual-studio/)
3) Pre-installed on the system via some tool (e.g. container image).

Under all of these scenarios, how should we expect tools to invoke `vcpkg`? The options as I see them:
1) `vcpkg` is on the PATH environment variable OR
2) Invocation should be through a rooted, non-relative path that is provided to the tool.
3) There exists some conventional environment variable, such as `VCPKG_ROOT`, at which `vcpkg` could be located.

For option 3), it would be unwise to have that conventional environment variable differ according to the technique that was used to obtain the vcpkg tool, which was the current case in the GitHub Actions containers, where `VCPKG_INSTALLATION_ROOT` was used instead of `VCPKG_ROOT`. Furthermore, any users who are techniques 1) or 2) to obtain vcpkg should be prepared to deal with systems where it already exists via technique 3), as it will be a very common way of obtaining vcpkg as soon as it begins being shipped more widely by other tools.

Another issue was identified where a user was using the presence of `VCPKG_ROOT` as an opt-in signal (see https://github.com/actions/runner-images/pull/6192#issuecomment-1249557653), which is the sort of situation of not having an agreed-upon convention has caused. Users should have a single environment variable whose sole purpose is to provide the location of the tool and should not be relied upon for other purposes (such as the aforementioned opt-in, unless it really is tied to the availbility of the tool).

------

**If we decide that we do not want to support option 3) for identifying the location of the tool, then we should make it clear that no one should rely upon such an environment variable being defined and that they should instead use options 1) or 2).**

**Comment by BillyONeal** [comment]
`VCPKG_ROOT` is the location of:

(1) the location of the built in registry, and
(2) the default root location for some of the other values, like the installed package root.

It defaults to the directory "above" the vcpkg binary with the file `.vcpkg-root`.

It was added because once upon a time, the canonical way to use vcpkg required building the vcpkg binary, and people asked for some way to avoid needing to rebuild the executable over and over, as that took significant time. So you could build / download the executable once, and reuse it with different "instances" of vcpkg.

In classic mode, it, in effect, controls the versions of all of the packages that one can install. A CI system setting this is, in effect, saying "to hell with the versions the project being built asked for, you are getting the system versions".

> There are at least 3 techniques being used in the wild to acquire vcpkg.
> 
> 1. Included as a git submodule, which is then acquired/updated using git's submodule commands.

In this case, the default `VCPKG_ROOT` will already be chosen correctly. `VCPKG_ROOT` being set by a CI system is unexpected and breaks these customers because they get incorrect versions of packages, and the packages are installed in the wrong place.

> 2. Downloaded via a one-line shell command (as shown in https://devblogs.microsoft.com/cppblog/vcpkg-environment-activation-in-visual-studio/)

In this case, vcpkg is being installed into the *shell* in experimental support of the "artifacts" feature, where vcpkg needs to be able to reach into the outer shell and set environment variables. `VCPKG_ROOT` will be set by that installation script to the extracted vcpkg binary, but this way of deploying vcpkg doesn't work in classic mode, so the problem of that variable controlling versions of everything is not a problem.

> 3. Pre-installed on the system via some tool (e.g. container image).

This looks like case 1.

In all 3 cases, the expected way the user calls us is via PATH (or by equivalent shell function in case 2).

> Under all of these scenarios, how should we expect tools to invoke `vcpkg`? The options as I see them:
> 
> 1. `vcpkg` is on the PATH environment variable OR
> 2. Invocation should be through a rooted, non-relative path that is provided to the tool.
> 3. There exists some conventional environment variable, such as `VCPKG_ROOT`, at which `vcpkg` could be located.

The vast vast majority of the time, it should be (1).

> For option 3), it would be unwise to have that conventional environment variable differ according to the technique that was used to obtain the vcpkg tool, which was the current case in the GitHub Actions containers, where `VCPKG_INSTALLATION_ROOT` was used instead of `VCPKG_ROOT`. Furthermore, any users who are techniques 1) or 2) to obtain vcpkg should be prepared to deal with systems where it already exists via technique 3), as it will be a very common way of obtaining vcpkg as soon as it begins being shipped more widely by other tools.

There is no such "conventional environment variable". I don't know if one is even possible because the different deployment mechanisms have intentionally different behavior; the "one liner" installer was added as an experiment and we need to take a long hard look at all of our configuration before taking the experimental tag off. The other methods reduce to PATH.

**Comment by BillyONeal** [comment]
We are still getting constant reports of people broken by this setting and still do not understand why GitHub Actions is setting it. We are also considering product changes to explicitly ignore the setting with a warning because CI systems keep doing this to us.

**Comment by BillyONeal** [comment]
Our workaround for this in the vcpkg product just landed today.

**Pull request merged**